torchtune/utils/_profiler.py

```diff
@@ -4,47 +4,350 @@
+    if len(activities) == 0:
+        _warn("No activities specified, defaulting to CPU + CUDA")
+        activities = _DEFAULT_PROFILER_ACTIVITIES
+        cpu = cuda = True
+
+    # Set up profiler schedule
+    use_default_schedule = not any([wait, warmup, active, repeat])
+
+    # Use default schedule if None, else validate that schedule is valid and can be passed to `instantiate`
+    if use_default_schedule:
+        wait = _DEFAULT_SCHEDULE["wait"]
+        warmup = _DEFAULT_SCHEDULE["warmup"]
+        active = _DEFAULT_SCHEDULE["active"]
+        repeat = _DEFAULT_SCHEDULE["repeat"]
+        _warn(
+            " No schedule found in config, defaulting to wait {wait}, warmup {warmup}, active {active}, repeat {repeat}"
         )
-        if enabled
-        else contextlib.nullcontext()
+    else:
+        if not all([wait, warmup, active]):
+            raise ValueError(
+                "Invalid schedule config: must specify wait, warmup, and active"
+            )
+        if repeat is None:
+            _warn(
+                """ No repeat found in schedule config, setting to 1 (one cycle).
+                If you want to cycle continuously, specify `repeat = 0`"""
+            )
+            repeat = 1
+
+    schedule = torch.profiler.schedule(
+        wait=wait, warmup=warmup, active=active, repeat=repeat
     )
+
+    # profile_memory requires with_stack and record_shapes, hence we override these if profile_memory is True
+    # See torch.profiler.profiler._memory_profile
+    if profile_memory:
+        _warn(
+            "`profile_memory` requires `with_stack` and `record_shapes`, these will be enabled since `profile_memory` is True"
+        )
+    with_stack = with_stack or profile_memory
+    record_shapes = record_shapes or profile_memory
+    # experimental config is needed to export stacks: see https://github.com/pytorch/pytorch/issues/100253
+    experimental_config = _ExperimentalConfig(verbose=True) if with_stack else None
+
+    # Handle exporting of trace, memory timeline and other profiler artifacts
+    if output_dir is None:
+        _warn(
+            f" No output directory found in profiler config, defaulting to {_DEFAULT_PROFILE_DIR}"
+        )
+        output_dir = _DEFAULT_PROFILE_DIR
+
+    output_dir = Path(output_dir)
+    output_dir.mkdir(parents=True, exist_ok=True)
+    output_dir = str(output_dir)
+
+    # trace_handler manages the export of profiler artifacts
+    # this callback will be triggered after **each** profiling cycle
+    callback = partial(trace_handler, output_dir=output_dir)
+
+    profiler = torch.profiler.profile(
+        activities=activities,
+        profile_memory=profile_memory,
+        with_stack=with_stack,
+        record_shapes=record_shapes,
+        with_flops=with_flops,
+        schedule=schedule,
+        experimental_config=experimental_config,
+        on_trace_ready=callback,
+    )
+
+    profiler_cfg = DictConfig(
```

Comment:
Sorry I actually disagree with @RdoubleA's [previous comment](https://github.com/pytorch/torchtune/pull/1089#discussion_r1639934356) here. I think it's better to keep the return type of `setup_torch_profiler` simple, I don't expect to get back a Tuple when calling `setup_torch_profiler`, I expect to just get back a profiler. Anyways, don't want to cause too much thrash on this so don't worry about changing it back for now; I think that's something we can sort out separately. But do lmk if you have any thoughts on this

Reply:
How about handle logging the config here instead of in the recipe? In fact you can just pretty print a regular duct instead of a DictConfig object. Then you don't need to return a Tuple.